File: checks/files.py
"""files — probe for exposed sensitive files and directory listings."""
import urllib.parse
from common import Finding

NAME = "files"

# path -> (severity, why, content marker that confirms exposure)
SENSITIVE = {
    "/.git/HEAD": ("high", "exposed .git repository", "ref:"),
    "/.git/config": ("high", "exposed git config", "[core]"),
    "/.env": ("critical", "exposed environment file (secrets)", "="),
    "/.env.local": ("critical", "exposed local env file", "="),
    "/config.json": ("medium", "exposed config", "{"),
    "/wp-config.php.bak": ("critical", "backup of WP config", "DB_PASSWORD"),
    "/backup.sql": ("critical", "exposed SQL dump", "INSERT INTO"),
    "/.DS_Store": ("low", "macOS directory metadata leak", "Bud1"),
    "/phpinfo.php": ("high", "phpinfo() exposed", "PHP Version"),
    "/server-status": ("medium", "Apache server-status exposed", "Server Version"),
    "/actuator/env": ("high", "Spring Actuator env exposed", "propertySources"),
    "/.well-known/security.txt": ("info", "security.txt present (good practice)", "Contact"),
}


def run(client, opts):
    out = []
    parsed = urllib.parse.urlparse(client.base_url)
    root = f"{parsed.scheme}://{parsed.netloc}"

    for path, (sev, why, marker) in SENSITIVE.items():
        r = client.request("GET", url=root + path)
        if r.status == 200 and marker.lower() in r.body.lower():
            out.append(Finding(NAME, sev, f"Exposed {path}", why, r.body[:120]))

    # Directory listing check on root
    r = client.request("GET", url=root + "/")
    if "Index of /" in r.body or "<title>Directory listing" in r.body:
        out.append(Finding(NAME, "medium", "Directory listing enabled",
                           "server returns an auto-index page", "Index of /"))
    return out

File: crates/dast/src/checks/files.rs
use crate::opts::Opts;
use crate::registry::CheckFuture;
use pentest_core::{Finding, HttpClient, HttpRequest, Severity};

pub const NAME: &str = "files";

const SENSITIVE: &[(&str, Severity, &str, &str)] = &[
    (
        "/.git/HEAD",
        Severity::High,
        "exposed .git repository",
        "ref:",
    ),
    (
        "/.git/config",
        Severity::High,
        "exposed git config",
        "[core]",
    ),
    (
        "/.env",
        Severity::Critical,
        "exposed environment file (secrets)",
        "=",
    ),
    (
        "/.env.local",
        Severity::Critical,
        "exposed local env file",
        "=",
    ),
    ("/config.json", Severity::Medium, "exposed config", "{"),
    (
        "/wp-config.php.bak",
        Severity::Critical,
        "backup of WP config",
        "DB_PASSWORD",
    ),
    (
        "/backup.sql",
        Severity::Critical,
        "exposed SQL dump",
        "INSERT INTO",
    ),
    (
        "/.DS_Store",
        Severity::Low,
        "macOS directory metadata leak",
        "Bud1",
    ),
    (
        "/phpinfo.php",
        Severity::High,
        "phpinfo() exposed",
        "PHP Version",
    ),
    (
        "/server-status",
        Severity::Medium,
        "Apache server-status exposed",
        "Server Version",
    ),
    (
        "/actuator/env",
        Severity::High,
        "Spring Actuator env exposed",
        "propertySources",
    ),
    (
        "/.well-known/security.txt",
        Severity::Info,
        "security.txt present (good practice)",
        "Contact",
    ),
];

pub fn run<'a>(client: &'a HttpClient, opts: &'a Opts) -> CheckFuture<'a> {
    Box::pin(run_impl(client, opts))
}

async fn run_impl(client: &HttpClient, _opts: &Opts) -> Vec<Finding> {
    let mut out = Vec::new();
    let root = client.base_url_root();

    for (path, sev, why, marker) in SENSITIVE {
        if let Ok(r) = client
            .request(HttpRequest::get().url(format!("{root}{path}")))
            .await
        {
            if r.status == 200 && r.body.to_lowercase().contains(&marker.to_lowercase()) {
                out.push(
                    Finding::new(NAME, *sev, format!("Exposed {path}"), *why)
                        .with_evidence(r.body.chars().take(120).collect::<String>()),
                );
            }
        }
    }

    if let Ok(r) = client
        .request(HttpRequest::get().url(format!("{root}/")))
        .await
    {
        if r.body.contains("Index of /") || r.body.contains("<title>Directory listing") {
            out.push(
                Finding::new(
                    NAME,
                    Severity::Medium,
                    "Directory listing enabled",
                    "server returns an auto-index page",
                )
                .with_evidence("Index of /".to_string()),
            );
        }
    }

    out
}
